components/scream/src/physics/p3/p3_inputs_initializer.cpp

```diff
@@ -45,88 +47,214 @@ add_field (const field_type &f, const field_type& f_ref,
+  view_2d d_exner("exner",ncol,nk_pack);          
   
-  // Call f90 routine
-  p3_standalone_init_f90 (q, T_atm, zi, pmid, dpres, ast, ni_activated, nc_nuceat_tend, qv_prev, t_prev);
+  // Create host mirrors for all views
+  auto h_T_atm            = Kokkos::create_mirror_view(d_T_atm          ); 
+  auto h_ast              = Kokkos::create_mirror_view(d_ast            ); 
+  auto h_ni_activated     = Kokkos::create_mirror_view(d_ni_activated   ); 
+  auto h_nc_nuceat_tend   = Kokkos::create_mirror_view(d_nc_nuceat_tend ); 
+  auto h_pmid             = Kokkos::create_mirror_view(d_pmid           ); 
+  auto h_dp               = Kokkos::create_mirror_view(d_dp             ); 
+  auto h_zi               = Kokkos::create_mirror_view(d_zi             ); 
+  auto h_qv_prev          = Kokkos::create_mirror_view(d_qv_prev        ); 
+  auto h_T_prev           = Kokkos::create_mirror_view(d_T_prev         ); 
+  auto h_qv               = Kokkos::create_mirror_view(d_qv             ); 
+  auto h_qc               = Kokkos::create_mirror_view(d_qc             ); 
+  auto h_qr               = Kokkos::create_mirror_view(d_qr             ); 
+  auto h_qi               = Kokkos::create_mirror_view(d_qi             ); 
+  auto h_qm               = Kokkos::create_mirror_view(d_qm             ); 
+  auto h_nc               = Kokkos::create_mirror_view(d_nc             ); 
+  auto h_nr               = Kokkos::create_mirror_view(d_nr             ); 
+  auto h_ni               = Kokkos::create_mirror_view(d_ni             ); 
+  auto h_bm               = Kokkos::create_mirror_view(d_bm             ); 
+  auto h_nccn_prescribed  = Kokkos::create_mirror_view(d_nccn_prescribed); 
+  auto h_inv_qc_relvar    = Kokkos::create_mirror_view(d_inv_qc_relvar  ); 
+  auto h_th_atm           = Kokkos::create_mirror_view(d_th_atm         ); 
+  auto h_dz               = Kokkos::create_mirror_view(d_dz             ); 
+  auto h_exner            = Kokkos::create_mirror_view(d_exner          ); 
 
-  // Deep copy back to device
-  Kokkos::deep_copy(d_q,h_q);
-  Kokkos::deep_copy(d_T,h_T);
-  Kokkos::deep_copy(d_ast,h_ast);
-  Kokkos::deep_copy(d_ni_activated,h_ni_activated);
-  Kokkos::deep_copy(d_nc_nuceat_tend,h_nc_nuceat_tend);
-  Kokkos::deep_copy(d_pmid,h_pmid);
-  Kokkos::deep_copy(d_dpres,h_dpres);
-  Kokkos::deep_copy(d_zi,h_zi);
-  Kokkos::deep_copy(d_qv_prev,h_qv_prev);
-  Kokkos::deep_copy(d_t_prev,h_t_prev);
-
-  // If we are in charge of init-ing FQ as well, init it to 0.
-  if (m_fields.count("FQ")==1) {
-    // Init FQ to 0
-    auto d_FQ = m_fields.at("FQ").get_view();
-    Kokkos::deep_copy(d_FQ,Real(0));
+ // Initalize from text file 
+  std::ifstream fid("p3_init_vals.txt", std::ifstream::in);
+  EKAT_REQUIRE_MSG(!fid.fail(),"Error in p3_inputs_initializer.cpp loading p3_init_vals.txt file");
+  std::string tmp_line;
+  getline(fid,tmp_line);  // Read header and discard.
+  int icol_in_max = 0;
+  while(getline(fid,tmp_line))
```

Comment:
Just a thought: using a file format that pairs field names with field data is more robust than assuming an ordering for the fields, since you can add and delete fields in the future without disrupting a sequence (and you can ignore any fields you don't need). Might not be worth fussing over, but I thought I'd mention it.

Reply:
I agree, in this case I was able to generate the input file directly from the `p3_run_and_cmp` output, which is why I kept it intact here, just in case there was some change in p3_run_and_cmp that I wanted reflected in the input file, and I didn't want to have to do any more formatting - i.e. lazy.

that being said, I didn't really put my heart into it because I think the more robust unit test will touch this code and would be the appropriate time to clean things up for good.